// @mui/material-next/node/CircularProgress/CircularProgress.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _clsx = _interopRequireDefault(require("clsx"));
var _utils = require("@mui/utils");
var _composeClasses = require("@mui/base/composeClasses");
var _system = require("@mui/system");
var _useThemeProps = _interopRequireDefault(require("../styles/useThemeProps"));
var _styled = _interopRequireDefault(require("../styles/styled"));
var _circularProgressClasses = require("./circularProgressClasses");
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["className", "color", "disableShrink", "fourColor", "style", "size", "thickness", "value", "variant"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const SIZE = 48;
const ARC_DURATION = 1333; // milliseconds
const CYCLE_DURATION = 4 * ARC_DURATION; // milliseconds
const LINEAR_ROTATE_DURATION = ARC_DURATION * 360 / 306; // milliseconds

const circularRotateKeyframe = (0, _system.keyframes)`
  to {
    transform: rotate(360deg);
  }
`;
const circularDashKeyframe = (0, _system.keyframes)`
  0% {
    stroke-dasharray: 10px, 200px;
    stroke-dashoffset: 0px;
  }

  50% {
    stroke-dasharray: 139px, 10px;
    stroke-dashoffset: -10px;
  }

  100% {
    stroke-dasharray: 139px, 129px;
    stroke-dashoffset: -139px;
  }
`;
const fourColorKeyframe = (0, _system.keyframes)`
  0% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-one-color);
  }
  15% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-one-color);
  }
  25% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-two-color);
  }
  40% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-two-color);
  }
  50% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-three-color);
  }
  65% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-three-color);
  }
  75% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-four-color);
  }
  90% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-four-color);
  }
  100% {
    stroke: var(--md-comp-linear-progress-indicator-four-color-active-indicator-one-color);
  }
`;
const useUtilityClasses = ownerState => {
  const {
    classes,
    variant,
    color,
    disableShrink,
    fourColor
  } = ownerState;
  const slots = {
    root: ['root', variant, `color${(0, _utils.unstable_capitalize)(color)}`, fourColor && 'fourColor', disableShrink && 'disableShrink'],
    svg: ['svg'],
    circle: ['circle']
  };
  return (0, _composeClasses.unstable_composeClasses)(slots, _circularProgressClasses.getCircularProgressUtilityClass, classes);
};
const CircularProgressRoot = (0, _styled.default)('span', {
  name: 'MuiCircularProgress',
  slot: 'Root',
  overridesResolver: (props, styles) => {
    const {
      ownerState
    } = props;
    return [styles.root, styles[ownerState.variant], styles[`color${(0, _utils.unstable_capitalize)(ownerState.color)}`], ownerState.fourColor && styles.fourColor, ownerState.disableShrink && styles.disableShrink];
  }
})(({
  ownerState,
  theme: {
    vars: tokens
  }
}) => (0, _extends2.default)({
  '--md-comp-linear-progress-indicator-active-indicator-color': ownerState.color !== 'inherit' ? tokens.sys.color[ownerState.color] : 'currentColor',
  '--md-comp-linear-progress-indicator-active-indicator-width': ownerState.thickness,
  '--md-comp-linear-progress-indicator-size': ownerState.size,
  '--md-comp-linear-progress-indicator-four-color-active-indicator-one-color': tokens.sys.color.primary,
  '--md-comp-linear-progress-indicator-four-color-active-indicator-two-color': tokens.sys.color.onPrimaryContainer,
  '--md-comp-linear-progress-indicator-four-color-active-indicator-three-color': tokens.sys.color.tertiary,
  '--md-comp-linear-progress-indicator-four-color-active-indicator-four-color': tokens.sys.color.onTertiaryContainer,
  display: 'inline-block',
  height: 'var(--md-comp-linear-progress-indicator-size)',
  width: 'var(--md-comp-linear-progress-indicator-size)'
}, ownerState.variant === 'determinate' && {
  transition: `transform ${tokens.sys.motion.duration.medium2} ${tokens.sys.motion.easing.legacy}`
}, ownerState.variant === 'indeterminate' && {
  animation: `linear infinite ${circularRotateKeyframe}`,
  animationDuration: `${LINEAR_ROTATE_DURATION}ms`
}));
const CircularProgressSVG = (0, _styled.default)('svg', {
  name: 'MuiCircularProgress',
  slot: 'Svg',
  overridesResolver: (props, styles) => styles.svg
})({
  display: 'block' // Keeps the progress centered
});
const CircularProgressCircle = (0, _styled.default)('circle', {
  name: 'MuiCircularProgress',
  slot: 'Circle',
  overridesResolver: (props, styles) => styles.circle
})(({
  ownerState,
  theme: {
    vars: tokens
  }
}) => (0, _extends2.default)({
  stroke: 'var(--md-comp-linear-progress-indicator-active-indicator-color)',
  strokeWidth: 'var(--md-comp-linear-progress-indicator-active-indicator-width)'
}, ownerState.variant === 'determinate' && {
  transition: `stroke-dashoffset ${tokens.sys.motion.duration.medium2} ${tokens.sys.motion.easing.legacy}`
}, ownerState.variant === 'indeterminate' && {
  // Some default value that looks fine waiting for the animation to kicks in.
  strokeDasharray: '10px, 200px',
  strokeDashoffset: 0 // Add the unit to fix a Edge 16 and below bug.
}, ownerState.variant === 'indeterminate' && !ownerState.disableShrink && (0, _extends2.default)({
  animation: circularDashKeyframe,
  animationDuration: `${ARC_DURATION}ms, ${CYCLE_DURATION}ms`,
  animationIterationCount: 'infinite',
  animationFillMode: 'both',
  animationTimingFunction: tokens.sys.motion.easing.legacy
}, ownerState.fourColor && {
  animationName: `${circularDashKeyframe}, ${fourColorKeyframe}`
})));

/**
 * ## ARIA
 *
 * If the progress bar is describing the loading progress of a particular region of a page,
 * you should use `aria-describedby` to point to the progress bar, and set the `aria-busy`
 * attribute to `true` on that region until it has finished loading.
 *
 * Demos:
 *
 * - [Progress](https://mui.com/material-ui/react-progress/)
 *
 * API:
 *
 * - [CircularProgress API](https://mui.com/material-ui/api/circular-progress/)
 */
const CircularProgress = /*#__PURE__*/React.forwardRef(function CircularProgress(inProps, ref) {
  const props = (0, _useThemeProps.default)({
    props: inProps,
    name: 'MuiCircularProgress'
  });
  const {
      className,
      color = 'primary',
      disableShrink = false,
      fourColor = false,
      style,
      size = 48,
      thickness = 4,
      value = 0,
      variant = 'indeterminate'
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const ownerState = (0, _extends2.default)({}, props, {
    color,
    disableShrink,
    fourColor,
    size: typeof size === 'number' ? `${size}px` : size,
    thickness,
    value,
    variant
  });
  const classes = useUtilityClasses(ownerState);
  const circleStyle = {};
  const rootStyle = {};
  const rootProps = {};
  if (variant === 'determinate') {
    const circumference = 2 * Math.PI * ((SIZE - thickness) / 2);
    circleStyle.strokeDasharray = circumference.toFixed(3);
    rootProps['aria-valuenow'] = Math.round(value);
    circleStyle.strokeDashoffset = `${((100 - value) / 100 * circumference).toFixed(3)}px`;
    rootStyle.transform = 'rotate(-90deg)';
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(CircularProgressRoot, (0, _extends2.default)({
    className: (0, _clsx.default)(classes.root, className),
    style: (0, _extends2.default)({}, rootStyle, style),
    ownerState: ownerState,
    ref: ref,
    role: "progressbar"
  }, rootProps, other, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(CircularProgressSVG, {
      className: classes.svg,
      ownerState: ownerState,
      viewBox: `${SIZE / 2} ${SIZE / 2} ${SIZE} ${SIZE}`,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(CircularProgressCircle, {
        className: classes.circle,
        style: circleStyle,
        ownerState: ownerState,
        cx: SIZE,
        cy: SIZE,
        r: (SIZE - thickness) / 2,
        fill: "none"
      })
    })
  }));
});
process.env.NODE_ENV !== "production" ? CircularProgress.propTypes /* remove-proptypes */ = {
  // ┌────────────────────────────── Warning ──────────────────────────────┐
  // │ These PropTypes are generated from the TypeScript type definitions. │
  // │ To update them, edit the TypeScript types and run `pnpm proptypes`. │
  // └─────────────────────────────────────────────────────────────────────┘
  /**
   * @ignore
   */
  children: _propTypes.default.node,
  /**
   * Override or extend the styles applied to the component.
   */
  classes: _propTypes.default.object,
  /**
   * @ignore
   */
  className: _propTypes.default.string,
  /**
   * The color of the component.
   * It supports both default and custom theme colors, which can be added as shown in the
   * [palette customization guide](https://mui.com/material-ui/customization/palette/#custom-colors).
   * @default 'primary'
   */
  color: _propTypes.default.oneOf(['error', 'info', 'inherit', 'primary', 'secondary', 'success', 'tertiary', 'warning']),
  /**
   * If `true`, the shrink animation is disabled.
   * This only works if variant is `indeterminate`.
   * @default false
   */
  disableShrink: (0, _utils.chainPropTypes)(_propTypes.default.bool, props => {
    if (props.disableShrink && props.variant && props.variant !== 'indeterminate') {
      return new Error('MUI: You have provided the `disableShrink` prop ' + 'with a variant other than `indeterminate`. This will have no effect.');
    }
    return null;
  }),
  /**
   * If `true`, the component render indeterminate mode using four colors instead of one.
   * This only works if variant is `indeterminate`.
   * @default false
   */
  fourColor: (0, _utils.chainPropTypes)(_propTypes.default.bool, props => {
    if (props.fourColor && props.variant && props.variant !== 'indeterminate') {
      return new Error('MUI: You have provided the `fourColor` prop ' + 'with a variant other than `indeterminate`. This will have no effect.');
    }
    return null;
  }),
  /**
   * The size of the component.
   * If using a number, the pixel unit is assumed.
   * If using a string, you need to provide the CSS unit, e.g. '3rem'.
   * @default 48
   */
  size: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  /**
   * @ignore
   */
  style: _propTypes.default.object,
  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.object, _propTypes.default.bool])), _propTypes.default.func, _propTypes.default.object]),
  /**
   * The thickness of the circle.
   * @default 4
   */
  thickness: _propTypes.default.number,
  /**
   * The value of the progress indicator for the determinate variant.
   * Value between 0 and 100.
   * @default 0
   */
  value: _propTypes.default.number,
  /**
   * The variant to use.
   * Use indeterminate when there is no progress value.
   * @default 'indeterminate'
   */
  variant: _propTypes.default.oneOf(['determinate', 'indeterminate'])
} : void 0;
var _default = exports.default = CircularProgress;
