// @mui/material-next/node/Tabs/TabsList.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _styles = require("@mui/material/styles");
var _useTabsList = require("@mui/base/useTabsList");
var _TabsListContext = _interopRequireDefault(require("./TabsListContext"));
var _jsxRuntime = require("react/jsx-runtime");
const _excluded = ["variant", "indicator", "textColor", "children"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const FlexContainer = (0, _styles.styled)('div', {
  name: 'MuiTabs',
  slot: 'FlexContainer',
  overridesResolver: (props, styles) => {
    const {
      ownerState
    } = props;
    return [styles.flexContainer, ownerState.vertical && styles.flexContainerVertical, ownerState.centered && styles.centered];
  }
})(({
  ownerState
}) => (0, _extends2.default)({
  display: 'flex'
}, ownerState.vertical && {
  flexDirection: 'column'
}, ownerState.centered && {
  justifyContent: 'center'
}));

/**
 * @ignore - internal component.
 */
const TabsList = /*#__PURE__*/React.forwardRef((props, ref) => {
  const {
      variant,
      indicator,
      textColor,
      children
    } = props,
    other = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  const {
    getRootProps,
    contextValue
  } = (0, _useTabsList.useTabsList)((0, _extends2.default)({}, props, {
    rootRef: ref
  }));
  const tabsListContextValue = React.useMemo(() => ({
    indicator,
    textColor,
    fullWidth: variant === 'fullWidth'
  }), [indicator, textColor, variant]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_useTabsList.TabsListProvider, {
    value: contextValue,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_TabsListContext.default.Provider, {
      value: tabsListContextValue,
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(FlexContainer, (0, _extends2.default)({}, other, getRootProps(), {
        children: children
      }))
    })
  });
});
process.env.NODE_ENV !== "production" ? TabsList.propTypes = {
  children: _propTypes.default.node,
  indicator: _propTypes.default.node,
  textColor: _propTypes.default.oneOf(['primary', 'secondary']),
  variant: _propTypes.default.oneOf(['fullWidth', 'scrollable', 'standard'])
} : void 0;
var _default = exports.default = TabsList;
